Allow renaming imports with `as`

PgBiel opened the pull request:
The `as` keyword has been reserved since the dawn of Typst, but never properly used. This PR implements renaming imports using the `as` keyword, thus giving it use.

# Highlights

**New syntax:**
1. `#import "module.typ" as something` (under `SyntaxKind::RenamedModuleImport`) now imports the module's entire scope under the variable name `something` instead of `module`
2. `#import "module.typ": func as dothing` (under `SyntaxKind::RenamedImport`) now imports the `func` name (defined in the module) as `dothing` in the current scope, instead of necessarily importing it as `func`

Note that, for consistency with the current behavior, I made it so you can't import the entire scope under a new name and, at the same time, import items from it (which generates the error below). This is, however, not a hard restriction and can be lifted.

```
error: unexpected colon
  ┌─ /aaa.typ:2:22
  │
2 │ #import "bbb.typ" as f: g
  │                       ^
  │
  = hint: cannot import the renamed module and its items at the same time
  = hint: try importing the module's items in a separate `import` expression
```

# Implementation details

- **New syntax node kinds:** `RenamedModuleImport` and `RenamedImport`
  - Parser changes weren't very complex
- **AST changes:**
  - Added `RenamedModuleImport` and `RenamedImport` as AST nodes
  - Added `Expr::RenamedImport` for `RenamedModuleImport` (can't have items)
  - New enum, `ImportedItem`, representing either a `Simple` item (just an `Ident`, so, no renaming), or a `Renamed` item (`RenamedImport` node, containing original name and new name definitions for later usage by eval)
    - This enum contains a `bound_name()` method which indicates the final name of the imported item in the scope it's being imported to (regardless of it being renamed or not), in contrast to `original_name()`, which will always correspond to the definition/exported name of the item.
  - In `ModuleImport` items: changed `Vec<Ident>` to `Vec<ImportedItem>`
- **Eval changes:**
  - Implemented `Eval` for ast::RenamedModuleImport (simply took the `ast::ModuleImport` code but used the new name straight away, without calling a second function, since there are no items).
  - **(Review would be a good idea here)** Changed a thing in `eval/func.rs` - apparently there's code to capture variables into a closure, so I made it capture the `bound_name()` for imported items. Please tell me if this was the wrong call, but, in principle, seems to be working
- **IDE changes:**
  - Kept the current behavior of suggesting items based on their exported names (i.e., renaming is never suggested, as that doesn't make sense).
- Added several tests.

laurmaedje commented:
I switched the AST from owning syntax nodes to borrowing them (makes AST operations more light-weight and faster), but it introduced a bunch of conflicts here. Sorry about that!

PgBiel commented:
> I switched the AST from owning syntax nodes to borrowing them (makes AST operations more light-weight and faster), but it introduced a bunch of conflicts here. Sorry about that!

Rebased :)

laurmaedje commented on crates/typst-syntax/src/ast.rs:
```diff
@@ -195,6 +195,8 @@ pub enum Expr<'a> {
     /// A module import: `import "utils.typ": a, b, c`.
     Import(ModuleImport<'a>),
+    /// A renamed module import (without items): `import "file.typ" as utils`
+    RenamedImport(RenamedModuleImport<'a>),
```
Unless there's a good reason, this shouldn't require a separate syntax kind. You can use the normal `ModuleImport`.

laurmaedje commented on crates/typst-syntax/src/ast.rs:
```diff
@@ -1990,6 +1994,24 @@ impl<'a> ModuleImport<'a> {
+    }
+
+    /// The name to save the module's items under.
+    pub fn new_name(self) -> Ident<'a> {
```
this would be moved to `ModuleImport` and return `Option<Ident<'a>>`. you can filter skip_while you don't find a `SyntaxKind::As` child and then find the first match. see `ShowRule` for a similar thing.

PgBiel replied:
Gotcha. I only kept it as a separate node because I didn't know how I'd detect a rename here. I'll try your suggestion.

PgBiel replied:
Ok, it seems to have worked, thanks! :D

laurmaedje reviewed (commented):
Didn't do a full review yet, but I noticed this already so I thought I'd leave a comment.

PgBiel commented on crates/typst/src/eval/mod.rs:
```diff
@@ -1692,30 +1692,47 @@ fn apply_imports<V: IntoValue>(
+                for item in items.items() {
+                    let original_ident = item.original_name();
+                    if let Some(value) = scope.get(&original_ident) {
+                        if let ast::ImportItem::Renamed(renamed_item) = &item {
+                            if renamed_item.original_name().as_str()
+                                == renamed_item.new_name().as_str()
+                            {
+                                vm.vt.tracer.warn(SourceDiagnostic::warning(
+                                    renamed_item.span(),
+                                    "renaming imported name to its own name",
+                                ));
+                            }
+                        }
```
I have to admit I'm not particularly proud of having this check here (upon eval) instead of in the parser or AST generator or something like that. Perhaps this could be moved elsewhere, but I'm not sure.

laurmaedje replied:
same goes for strong and emph emptyness. but since there are no parser warnings atm, there's not much that can be done until we have those.

laurmaedje commented on crates/typst-syntax/src/ast.rs:
```diff
@@ -2006,8 +2015,67 @@ node! {
-        self.0.children().filter_map(SyntaxNode::cast)
+    pub fn items(self) -> impl DoubleEndedIterator<Item = ImportItem<'a>> {
+        self.0.children().filter_map(|child| {
+            child
+                .cast::<RenamedImportItem>()
+                .map(ImportItem::Renamed)
+                .or_else(|| child.cast::<Ident>().map(ImportItem::Simple))
```
I think the classical match on `child.kind()` is cleaner and easier to follow

laurmaedje commented on crates/typst-syntax/src/ast.rs:
```diff
@@ -2006,8 +2015,67 @@ node! {
+}
+
+impl<'a> ImportItem<'a> {
+    /// The original name of the imported item,
```
could you wrap comments at 80 cols?

PgBiel replied:
sorry, missed it here lol.

(Fun fact: I believe `rustfmt` has a built-in parameter that handles this, but AFAIK you need to use nightly to be able to pass parameters to rustfmt. ;-;)

laurmaedje commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -1138,18 +1138,40 @@ fn module_import(p: &mut Parser) {
-    if p.eat_if(SyntaxKind::Colon) && !p.eat_if(SyntaxKind::Star) {
-        import_items(p);
+    if p.eat_if(SyntaxKind::As) {
+        // allow renaming a full module import
```
could you capitalize the first word and end with a period?

laurmaedje commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -1138,18 +1138,40 @@ fn module_import(p: &mut Parser) {
+        if p.at(SyntaxKind::Colon) {
+            p.unexpected();
+            p.hint("cannot import the renamed module and its items at the same time");
+            p.hint("try importing the module's items in a separate `import` expression");
```
maybe still call `import_items` here to prevent further errors for them?

PgBiel replied:
I think that could be weird, mostly because that would make the parser check if the items are valid, when this is supposed to be invalid syntax, so it would say something like "missing identifier" in some item for example when, in reality, the entire construction is supposed to be invalid (not to mention that the marker wrap, if delayed as you suggest, would also include the items, which would make even less sense for an invalid construction!).

I can add this either way if you think that's fine though.

Of course, we could also just support importing the module and its items at the same time through renaming. I honestly don't have a very strong opinion here (I made it invalid only to keep the status quo for now, as it could then be added in a future PR if we ever change our minds later). I'll be keeping such behavior invalid for now, but feel free to tell me if you have diverging opinions in this sense.

PgBiel replied:
Update: It seems that calling `import_items` here is necessary to move the marker outside the 'if' as you suggest (otherwise a more general parsing error takes priority over the colon error). I'm not sure what I prefer, but if this seems like the best path to you then I'll push it.

laurmaedje replied:
> not to mention that the marker wrap, if delayed as you suggest, would also include the items, which would make even less sense for an invalid construction

I don't think this would make less sense actually. This is clearly how any human parses it and the parser should parse invalid constructions as closely to have humans interpret them, even if they don't follow the AST structure.

laurmaedje replied:
> Of course, we could also just support importing the module and its items at the same time through renaming. I honestly don't have a very strong opinion here (I made it invalid only to keep the status quo for now, as it could then be added in a future PR if we ever change our minds later). I'll be keeping such behavior invalid for now, but feel free to tell me if you have diverging opinions in this sense.

For a rename it makes sort of sense to allow this. I doesn't happen by default for non-renamed imports because there'd be no way to import items without importing the module then.

PgBiel replied:
> > not to mention that the marker wrap, if delayed as you suggest, would also include the items, which would make even less sense for an invalid construction
> 
> I don't think this would make less sense actually. This is clearly how any human parses it and the parser should parse invalid constructions as closely to how humans interpret them, even if they don't follow the AST structure.

Alright. I will be pushing the changes then. :+1:

PgBiel replied:
> > Of course, we could also just support importing the module and its items at the same time through renaming. I honestly don't have a very strong opinion here (I made it invalid only to keep the status quo for now, as it could then be added in a future PR if we ever change our minds later). I'll be keeping such behavior invalid for now, but feel free to tell me if you have diverging opinions in this sense.
> 
> For a rename it makes sort of sense to allow this. I doesn't happen by default for non-renamed imports because there'd be no way to import items without importing the module then.

I guess that makes sense (Rust also lets you do something similar). Although maybe it might seem to the end user that one line is doing "too much" (like `#import "something.typ" as x: y, z as b, w`). I'm not sure if that's even a real concern, though (maybe I'm just thinking too harshly). Do you think I should implement it then (should be easy)?

laurmaedje replied:
Why not

PgBiel replied:
Ok, done! :+1:

laurmaedje commented on crates/typst-syntax/src/parser.rs:
```diff
@@ -1138,18 +1138,40 @@ fn module_import(p: &mut Parser) {
+        }
+        p.wrap(m, SyntaxKind::ModuleImport);
     }
-    p.wrap(m, SyntaxKind::ModuleImport);
```
I think it's fine to do the wrapping here (for both cases) and the use `else if` above. it doesn't matter if the colon (and potentially the import items if my suggestion from above is applied) are part of the import if it's malformed. in fact, one could say that it makes more sense.

laurmaedje commented on crates/typst/src/eval/func.rs:
```diff
@@ -545,8 +545,8 @@ impl<'a> CapturesVisitor<'a> {
                 if let Some(ast::Imports::Items(items)) = expr.imports() {
-                    for item in items.idents() {
-                        self.bind(item);
+                    for item in items.items() {
```
I generally dislike if there's a variable and we access a field or call a method with its exact name. It feels like something wrong is going on. Maybe it's ok here, but maybe we should rename `items` to `iter` since the container itself is already `items`.

PgBiel replied:
`iter` definitely sounds nice.

laurmaedje commented on crates/typst/src/eval/mod.rs:
```diff
@@ -1692,30 +1692,47 @@ fn apply_imports<V: IntoValue>(
-                vm.scopes.top.define(var.clone(), value.clone());
+    if let Some(new_name) = new_name {
+        // Renamed module => just define it on the scope (no items)
+        vm.scopes.top.define(new_name, source_value);
```
early return instead of else?

laurmaedje commented on crates/typst/src/eval/mod.rs:
```diff
@@ -1692,30 +1692,47 @@ fn apply_imports<V: IntoValue>(
+                            if renamed_item.original_name().as_str()
+                                == renamed_item.new_name().as_str()
+                            {
+                                vm.vt.tracer.warn(SourceDiagnostic::warning(
```
it doesn't gain you much, but I'd still you the warning macro

PgBiel replied:
yeah it just took off a word haha, but it's fine

laurmaedje commented on tests/typ/compiler/import.typ:
```diff
@@ -34,6 +34,16 @@
+#import "module.typ": item as something
+#test(something(1, 2), 3)
+
+// Mixing up renamed and not renamed items.
```
```suggestion
// Mixing renamed and not renamed items.
```

laurmaedje commented on tests/typ/compiler/import.typ:
```diff
@@ -78,16 +111,27 @@
 #item(2)[a]
 
+---
+// Warning: 15-27 renaming imported name to its own name
```
I think the warning message could somehow be nicer. Maybe `unnecessary rename to same name`?

PgBiel replied:
I like that, maybe `unnecessary import rename to same name`?

laurmaedje commented:
Nice, pretty simple change in the end, especially in the parser!

laurmaedje commented:
Thank you!
